File: Medium/106. Construct Binary Tree from Inorder and Postorder Traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    TreeNode* buildTreeHelper(int &index, int left, int right, vector<int> &postorder, unordered_map<int, int> &inorderIndex){
        if (left > right)   return NULL;
        
        int curVal = postorder[index--];
        
        TreeNode* curNode = new TreeNode(curVal);
        
        curNode -> right = buildTreeHelper(index, inorderIndex[curVal] + 1, right, postorder, inorderIndex);
        curNode -> left = buildTreeHelper(index, left, inorderIndex[curVal] - 1, postorder, inorderIndex);
        
        return curNode;
    }
    
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        unordered_map<int, int> inorderIndex;
        
        for (int i = 0; i < inorder.size(); i++){
            inorderIndex[inorder[i]] = i;
        }
        
        int postorderIndex = postorder.size() - 1;
        
        return buildTreeHelper(postorderIndex, 0, inorder.size() - 1, postorder, inorderIndex);
    }
};

File: Medium/106. Construct Binary Tree from Inorder and Postorder Traversal/solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    
    //private Dictionary<int, int> hashMap;
    
    public TreeNode BuildTreeHelper(ref int[] inorder, ref int[] postorder, 
                                    ref Dictionary <int, int> inorderIndex, int left, int right, ref int index){
        
        
        if (left > right)   return null;
        
        
        TreeNode currentNode = new TreeNode(postorder[index]);
        int currentIndex = inorderIndex[postorder[index]];
        
        index--;
        
        //if (left == right)  return currentNode;
        
        
        //Console.WriteLine(index + " " + left + " " + right + " " + currentIndex);
        
        currentNode.right = BuildTreeHelper(ref inorder, ref postorder, ref inorderIndex, currentIndex + 1, right, ref index);
        currentNode.left = BuildTreeHelper(ref inorder, ref postorder, ref inorderIndex, left, currentIndex - 1, ref index);
        
        return currentNode;
    }
    
    public TreeNode BuildTree(int[] inorder, int[] postorder) {
        Dictionary <int, int> inorderIndex = new Dictionary<int, int> ();
        
        int length = inorder.Length;
        
        for (int i = 0; i < length; i++){
            inorderIndex[inorder[i]] = i;
        }
        
        int index = length - 1;
        
        return BuildTreeHelper(ref inorder, ref postorder, ref inorderIndex, 0, length - 1, ref index);
    }
}